File: core/src/main/java/com/mcreater/amclcore/model/oauth/session/MinecraftProfileRequestModel.java
package com.mcreater.amclcore.model.oauth.session;

import com.mcreater.amclcore.annotations.RequestModel;
import lombok.Builder;
import lombok.Data;

import java.util.List;
import java.util.Map;
import java.util.UUID;

@Builder
@Data
@RequestModel
public class MinecraftProfileRequestModel {
    private UUID id;
    private String name;
    private List<MinecraftProfileSkinModel> skins;
    private List<MinecraftProfileCapeModel> capes;
    private Map<Object, Object> profileActions;
    // if exception
    private String path;
    private String errorType;
    private String error;
    private MinecraftNameChangeableRequestModel details;
    private String errorMessage;
    private String developerMessage;

    @Data
    @Builder
    @RequestModel
    public static class MinecraftProfileSkinModel {
        private UUID id;
        private State state;
        private String url;
        private Variant variant;
    }

    @Data
    @Builder
    @RequestModel
    public static class MinecraftProfileCapeModel {
        private UUID id;
        private State state;
        private String url;
        private String alias;

        public String createId() {
            return id.toString();
        }
    }

    public enum State {
        ACTIVE,
        INACTIVE;

        public static State parse(String s) {
            try {
                return valueOf(s);
            } catch (Exception e) {
                return INACTIVE;
            }
        }
    }

    public enum Variant {
        CLASSIC,
        SLIM;

        public static Variant parse(String s) {
            try {
                return valueOf(s);
            } catch (Exception e) {
                return CLASSIC;
            }
        }
    }
}

File: core/src/main/java/com/mcreater/amclcore/model/oauth/session/MinecraftProfileRequestModel.kt
package com.mcreater.amclcore.model.oauth.session

import com.mcreater.amclcore.annotations.RequestModel
import java.util.*

@RequestModel
data class MinecraftProfileRequestModel(
    var id: UUID? = null,
    var name: String? = null,
    var skins: List<MinecraftProfileSkinModel>? = null,
    var capes: List<MinecraftProfileCapeModel>? = null,
    var profileActions: Map<Any?, Any?>? = null,
    var path: String? = null,
    var errorType: String? = null,
    var error: String? = null,
    var details: MinecraftNameChangeableRequestModel? = null,
    var errorMessage: String? = null,
    var developerMessage: String? = null
) {
    @RequestModel
    data class MinecraftProfileSkinModel(
        var id: UUID? = null,
        var state: State? = null,
        var url: String? = null,
        var variant: Variant? = null
    )


    @RequestModel
    data class MinecraftProfileCapeModel(
        var id: UUID? = null,
        var state: State? = null,
        var url: String? = null,
        var alias: String? = null
    ) {
        fun createId(): String {
            return id.toString()
        }
    }


    enum class State {
        ACTIVE,
        INACTIVE;

        companion object {
            @JvmStatic
            fun parse(s: String?): State {
                return try {
                    valueOf(s!!)
                } catch (e: Exception) {
                    INACTIVE
                }
            }
        }
    }


    enum class Variant {
        CLASSIC,
        SLIM;

        companion object {
            @JvmStatic
            fun parse(s: String?): Variant {
                return try {
                    valueOf(s!!)
                } catch (e: Exception) {
                    CLASSIC
                }
            }
        }
    }

}